File: src/nkit/gui/progress.nim
import ../foundation/event_emitter
import ./view

when defined(ios):
  import ../platform/ios/uifunctions
elif defined(macosx):
  import ../platform/macos/nsfunctions
elif defined(linux):
  import ../platform/linux/gfunctions

export view

type
  ProgressStyle* = enum
    psBar
    psSpinner

  Progress* = ref object of View

proc newProgress*(style: ProgressStyle = psBar, value = 0.0): Progress =
  when defined(macosx) or defined(ios) or defined(linux):
    let nativePtr = naProgressCreate(cint(ord(style)))
  else:
    let nativePtr: pointer = nil
  result = Progress()
  discard wrapView(result, nativePtr)
  when defined(macosx) or defined(ios) or defined(linux):
    if style == psBar and value != 0.0:
      naProgressSetValue(result.native, value)

proc destroy*(p: Progress) =
  when defined(macosx) or defined(ios) or defined(linux):
    naProgressFree(p.native)
    p.native = nil
  shutdownEmitter[GuiEvent](p)

proc setValue*(p: Progress, value: float64) =
  when defined(macosx) or defined(ios) or defined(linux):
    naProgressSetValue(p.native, value)

proc getValue*(p: Progress): float64 =
  when defined(macosx) or defined(ios) or defined(linux):
    naProgressGetValue(p.native)
  else:
    0.0

proc setIndeterminate*(p: Progress, indeterminate: bool) =
  when defined(macosx) or defined(ios) or defined(linux):
    naProgressSetIndeterminate(p.native, indeterminate)

proc isIndeterminate*(p: Progress): bool =
  when defined(macosx) or defined(ios) or defined(linux):
    naProgressIsIndeterminate(p.native)
  else:
    false

File: src/nkit/platform/ios/widgets/progress.m
#import "gui_common.h"

void* na_progress_create(int style) {
  UIProgressView* pv = [[UIProgressView alloc]
      initWithProgressViewStyle:(style == 1 ?
          UIProgressViewStyleDefault : UIProgressViewStyleBar)];
  pv.progress = 0.0;
  return (__bridge_retained void*)pv;
}

void na_progress_free(void* ptr) {
  if (!ptr) return;
  UIProgressView* pv = (__bridge_transfer UIProgressView*)ptr;
  [pv removeFromSuperview];
}

void na_progress_set_value(void* ptr, double value) {
  UIProgressView* pv = (__bridge UIProgressView*)ptr;
  if (pv) pv.progress = (float)value;
}

double na_progress_get_value(void* ptr) {
  UIProgressView* pv = (__bridge UIProgressView*)ptr;
  return pv ? (double)pv.progress : 0.0;
}

void na_progress_set_indeterminate(void* ptr, bool indeterminate) {
  (void)ptr; (void)indeterminate;
}

bool na_progress_is_indeterminate(void* ptr) {
  (void)ptr; return false;
}
